doc/sphinx/userguide/extensible-reactor.md

```diff
@@ -0,0 +1,156 @@
+     + \sum_{in} \dot{m}_{in} \left(h_{in} - \sum_k h_k Y_{k,in} \right)
+$$
+
+$$
+\frac{d(mY_k)}{dt} = \sum_{in} \dot{m}_{in} Y_{k,in} - \sum_{out} \dot{m}_{out} Y_k
+                     + \dot{m}_{k,gen}
+$$
+
+Each of these equations is written with an expression on the left hand side (LHS)
+multiplied by the time derivative of a state variable which equals an expression on the
+right hand side (RHS). The interface of the reactor network model is defined so that the
+reactor governing equations are evaluated by calculating these LHS and RHS expressions
+for each governing equation. The extensible reactor model then allows calculation of
+these LHS and RHS terms to be replaced or augmented by user-provided methods.
+
+For example, to add a term for a large mass (say a rock) inside the reactor that affects
+the thermal mass, the energy equation would become:
+
+$$
+\left(m c_p + m_{rock} c_{p,rock}\right) \frac{dT}{dt} = \dot{Q}
+    - \sum_k h_k \dot{m}_{k,gen}
+    + \sum_{in} \dot{m}_{in} \left(h_{in} - \sum_k h_k Y_{k,in} \right)
+$$
+
+Here, the LHS coefficient has changed from $m c_p$ to $m c_p + m_{rock} c_{p,rock}$. The
+other governing equations defining the ideal gas constant pressure reactor can be left
+unmodified. To implement this change, we define a new class derived from
+{py:class}`ExtensibleIdealGasConstPressureReactor`. In this new class, we then define an
+implementation for the `after_eval` method which has the signature
+
+```py
+def after_eval(self, t, LHS, RHS) -> None
+```
+
+where `t` is the current simulation time and `LHS` and `RHS` are arrays with one element
+for each state variable (in this case, two plus the number of species).
+
+Optional user-defined methods for the `ExtensibleReactor` class start with the prefix
+`before_`, `after_`, or `replace_`, indicating whether the this method should be called
+before, after, or instead of the corresponding method from the base class. In this case,
+our method will be called after the `eval` method of {ct}`IdealGasConstPressureReactor`,
+with the values of `LHS` and `RHS` already set based on the original governing
+equations. Instead of `after_eval`, we could implement either `before_eval` or
+`replace_eval` methods. However, in the case of `before_eval`, any element of `LHS` that
+we set would be overwritten by the initial implementation, and if we use `replace_eval`,
+we would have needed to calculate values for all elements of `RHS` and `LHS` instead of
+just modifying a single value.
+
+To obtain the above modification to the governing equations, the implementation of
+`after_eval` inside a new reactor class is then simply:
+
+```{code-cell} python
+import cantera as ct
+
+class RockReactor(ct.ExtensibleIdealGasConstPressureReactor):
+    mass_rock = 3  # kg
+    cp_rock = 790  # J/kg/K
+
+    def after_eval(self, t, LHS, RHS):
+        # The index 1 corresponds to the governing equation for the second
+        # state variable (temperature)
```

Comment:
At best, the Numpy structured array concept would only be a partial solution, since the indices into the RHS/LHS arrays also applies in the C++ layer, and to the ordering of other arrays like sensitivity coefficients.

Reply:
Yep, that makes sense. However, in the translation layer from C++ <-> Python where these functions get called, we could probably cast to/from structured arrays, right? At least that would help on the Python side, and changes to the C++ side would be handled internally, so the structured array fields would still be accessible.